tools/schemapi/utils.py

```diff
@@ -362,6 +365,16 @@ def __setattr__(self, name: str, value: Any) -> Never:
         msg = f"{type(self).__name__!r} is immutable.\nCould not assign self.{name} = {value}"
         raise TypeError(msg)
 
+    def __hash__(self) -> int:
+        return hash(_HASH_ENCODER.encode(self.schema))
+
+    def __eq__(self, value: object) -> bool:
+        if isinstance(value, SchemaInfo):
+            if self.ref:
+                return self.ref == value.ref
+            return hash(self) == hash(value)
+        return False
```

Comment:
# TODO
May be able to avoid the `hash` branch, with 

```py
class SchemaInfo:
	def __eq__(self, value: object) -> bool:
		if isinstance(value, SchemaInfo):
			if self.ref:
				return self.ref == value.ref
			return self.schema == value.schema
		return False
```

https://docs.python.org/3.8/library/stdtypes.html#dict.values
> Dictionaries compare equal if and only if they have the same `(key, value)` pairs (regardless of ordering). 
> Order comparisons (‘<’, ‘<=’, ‘>=’, ‘>’) raise [TypeError](https://docs.python.org/3.8/library/exceptions.html#TypeError).

Not sure if `dict.__eq__` short-circuits on the first non-eq pair, but could write that in if needed

Reply:
Fixed in [`a70f259` (#3536)](https://github.com/vega/altair/pull/3536/commits/a70f259f0e711fbaa69205976ed5a04e1633611c)